urlc5fdb179cc_lagodiuk_evo_neural_network_agents_tgz-pJ8-com_lagodiuk_agent_evolution_MainJ8/src/com/lagodiuk/agent/evolution/Main.java

```diff
@@ -0,0 +1,604 @@
+			}
+		});
+	}
+
+	private static void initializeResetButtonFunctionality() {
+		resetButton.addActionListener(new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent arg0) {
+				disableControls();
+
+				int populationSize = ga.getPopulation().getSize();
+				int parentalChromosomesSurviveCount = ga.getParentChromosomesSurviveCount();
+				initializeGeneticAlgorithm(populationSize, parentalChromosomesSurviveCount, null);
+
+				NeuralNetwork newBrain = ga.getBest();
+
+				setAgentBrains(newBrain);
+
+				// reset population number counter
+				populationNumber = 0;
+				populationInfoLabel.setText("Population: " + populationNumber);
+
+				enableControls();
+			}
+		});
+	}
+
+	private static void initializeEvolveButtonFunctionality() {
+		evolveButton.addActionListener(new ActionListener() {
+			@Override
+			public void actionPerformed(ActionEvent arg0) {
+				disableControls();
+				progressBar.setVisible(true);
+				progressBar.setValue(0);
+				environmentPanel.getGraphics().drawImage(displayEnvironmentBufferedImage, 0, 0, null);
+
+				String iterCountStr = evolveTextField.getText();
+				if (!iterCountStr.matches("\\d+")) {
+					evolveButton.setEnabled(true);
+					evolveTextField.setEnabled(true);
+					progressBar.setVisible(false);
+					environmentPanel.getGraphics().drawImage(displayEnvironmentBufferedImage, 0, 0, null);
+					return;
+				}
+
+				final int iterCount = Integer.parseInt(iterCountStr);
+
+				new Thread(new Runnable() {
+					@Override
+					public void run() {
+						IterartionListener<OptimizableNeuralNetwork, Double> progressListener =
+								new IterartionListener<OptimizableNeuralNetwork, Double>() {
+									@Override
+									public void update(GeneticAlgorithm<OptimizableNeuralNetwork, Double> environment) {
+										final int iteration = environment.getIteration();
+										SwingUtilities.invokeLater(new Runnable() {
+											@Override
+											public void run() {
+												progressBar.setValue((iteration * 100) / iterCount);
+											}
+										});
+									}
+								};
+
+						ga.addIterationListener(progressListener);
+						ga.evolve(iterCount);
+						ga.removeIterationListener(progressListener);
+						populationNumber += iterCount;
+
+						NeuralNetwork newBrain = ga.getBest();
+						setAgentBrains(newBrain);
+
+						SwingUtilities.invokeLater(new Runnable() {
+							@Override
+							public void run() {
+								progressBar.setVisible(false);
+								populationInfoLabel.setText("Population: " + populationNumber);
+								enableControls();
+								evolveButton.requestFocusInWindow();
+							}
+						});
+					}
+				}).start();
+			}
+		});
+	}
+
+	private static void disableControls() {
+		evolveButton.setEnabled(false);
+		evolveTextField.setEnabled(false);
+		loadBrainButton.setEnabled(false);
+		saveBrainButton.setEnabled(false);
+		staticFoodRadioButton.setEnabled(false);
+		dynamicFoodRadioButton.setEnabled(false);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 478:483, 487:492.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.